### PR TITLE
Optimize player vision, part 2

### LINKED ISSUE #4253: Player vision code is very slow
And I mean **very** slow. Profiling shows it is the biggest CPU eater right now by far.

Here's the problem. Every time a bob moves, its entire view area from its old position is recalculated:
https://github.com/widelands/widelands/blob/ff36ede854ab98cab34e20d6b2c2a3002c38a27b/src/logic/map_objects/bob.cc#L929-L934
That means every field separately:
https://github.com/widelands/widelands/blob/ff36ede854ab98cab34e20d6b2c2a3002c38a27b/src/logic/player.cc#L1486-L1489
And for each of these fields, iterate through all of our+allied buildings+bobs, and calculate if this field is in their field of vision:
https://github.com/widelands/widelands/blob/ff36ede854ab98cab34e20d6b2c2a3002c38a27b/src/logic/player.cc#L1419-L1445

That gives us up to `(moving bobs) * (π * radius^2) * (bobs+buildings)` vision checks. Oops!

**Potential solutions:**

1. Cache the number of object that see each field so there's no need to calculate it all the time. Basically what was done before 02947d690dae1f87131a6db4e686c047939cbdab (#4046).
2. Reduce the number of checked fields by not checking the fields that are seen from the moving bob's new position.
3. When checking a group of nearby fields and finding that a seer sees one of them, check that seer first in subsequent iterations.

**ADDED:**

4. When evaluating an area, do an initial iteration calculating the distance of all seers to the center field and put the ones that are close into a list. Then for the other fields only check seers from that list.
5. Divide the map into sectors. When checking visibility, only consider seers from the same and adjacent sectors.
6. Iterate seers belonging to the field's owner first.

I think option 1 is the only real solution. The other options could give good gains in an average case but would still leave big lag spikes in some scenarios.

**Version:** ff36ede854

### PR DESCRIPTION
**Pull request by Niektory**
A.k.a. The Real Deal.

Counting the number of seers for each field is back, with some twists. The vision values are:
0 = unexplored (never seen)
1 = explored/previously seen (not seen now)
\>1 = currently seen
2 = not seen by us, but seen by a team member
\>2 = seen by us; each seer adds +2, the revealed state (via Lua `reveal_fields()`) adds +1; so:
3 = revealed, no seers
4 = not revealed, 1 seer
5 = revealed, 1 seer
6 = not revealed, 2 seers
and so on.

Each team member counts only their own seers, making team splitting possible. We don't care about team vision if we can see the field ourselves. Encoding the revealed state in the counter saves memory when a lot of fields are revealed. Performance is comparable to Build 21. In other words, fast :-)

Fix #4253

### CODE REVIEW

**Comment by Niektory** [comment]
OK, this is mostly ready and testable. Saveloading might still need changes (it saves more than strictly necessary), but it should work. Savegame compatibility is not done though (how many versions do I need to support?).

**Comment by Noordfrees** [comment]
The packet was newly introduced after b21, so breaking compatibility affects only development versions. If it is possible without too much effort to provide compatibility, please do so, but if not it's not the end of the world.
After a quick look at the diff – please don't reinstate the numeric vision comparisons all over the code. If possible *all* comparisons should take place in dedicated functions in player.cc. Will test this soon.

**Comment by Niektory** [comment]
Should there be two vision types – numeric and enum – with conversions when appropriate, or just the numeric type with more helper functions?

**Comment by Noordfrees** [comment]
Perhaps an enum with types like `kUnexplored, kPreviouslySeen, kVisible` like now (and other constants if you need them) and conversion functions from the raw numeric vision to the enum types would be best. If the use of the vision number doesn't reasonably fit into such an enum, just more helper functions are fine.

**Comment by Niektory** [comment]
I ended up doing both, so `Vision` got some helper functions and is also convertible to the enum. `vision.is_visible()` and `vision == VisibleState::kVisible` are equivalent, I tend to use whatever is shorter. And in some places the enum type is used directly because only the 3 states are relevant.

**Comment by Niektory** [comment]
Savegame compatibility is in. Packet versions 4 and 3 work. 2 and 1 I couldn't test because of #4286.

"See all" mode is currently not working properly. Anything I should keep in mind when fixing it?

File: src/logic/hide_or_reveal_field_mode.h [hunk 1]
```diff
@@ -17,15 +17,10 @@
...
+#define WL_LOGIC_HIDE_OR_REVEAL_FIELD_MODE_H
 
 namespace Widelands {
-// Do not change the order! It is stored in savegames.
-enum class SeeUnseeNode {
-	kUnexplored = 0,      // Never seen yet
-	kPreviouslySeen = 1,  // Previously seen
-	kVisible = 2          // Currently visible
-};
+enum class HideOrRevealFieldMode { kHide, kHideAndForget, kReveal };
```
Comment (Noordfrees) [inline src/logic/hide_or_reveal_field_mode.h]: Move the enum to `vision.h` and delete this file

File: src/logic/vision.h [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+enum class VisibleState {
+	kUnexplored = 0,      // Never seen yet
+	kPreviouslySeen = 1,  // Previously seen
+	kVisible = 2          // Currently visible
+};
+
+// Extended vision information encoded as a number with some helper functions.
+// See Player::Field for explanation of values.
+struct Vision {
+	uint16_t value;
```
Comment (Noordfrees) [inline src/logic/vision.h]: Should this value be accessed directly? Better make it `private`, add a getter function, and either some functions or a `friend` for modifications.

**Review by Noordfrees — COMMENTED** [review]
Found no problems while testing so far. Code LGTM, a few comments inline.

**Review by Noordfrees — APPROVED** [review]
LGTM, no problems observed

**Comment by gunchleoc** [comment]
We have merge conflicts now

**Pull request merged**